compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -640,6 +640,17 @@ impl<'a> Parser<'a> {
             }
         }
 
+        // extra info for `c"str"` before 2021 edition or `c "str"` in all editions
+        if self.prev_token.is_ident_named(Symbol::intern("c"))
+            && let TokenKind::Literal(token::Lit { kind: token::Str | token::StrRaw(..), .. }) =
+                &self.token.kind
```

Comment:
You could check if the `c` immediately precedes the string literal I guess via `self.prev_token.span.hi() == self.token.span.lo()` to rule out the false positives like `c ""` (notice the space).

Reply:
While you noted

> We could probably figure out exactly which scenario has been encountered by examining spans and editions, but I figured it would be better not to overcomplicate the creation of the error too much.

it should be as simple as `self.prev_token.span.eq_ctxt(self.token.span) && self.token.span.edition < Edition::Edition2021`.

The `eq_ctxt` check prevents the notes about cstr literals from misfiring on:

```rs
// edition: <2021
macro_rules! m { ($x:ident) => { $x"" } }
fn main() { m!(c) }
```

Your current implementation fires here.

Reply:
I'm fixing this up, but using this macro as an example, `self.prev_token.span.eq_ctxt(self.token.span) == true`. The previous token's span appears to be that of the `$x` token in the macro expansion body, rather than the `c` argument to the macro, so they have the same context. Is this correct, or am I missing something?